doc/api/n-api.md

```diff
@@ -0,0 +1,3151 @@
+```
+and the format of the `napi_extended_error_info` structure is as follows:
+
+```C
+struct napi_extended_error_info {
+  const char* error_message;
+  void* engine_reserved;
+  uint32_t engine_error_code;
+  napi_status error_code;
+};
+```
+
+`napi_get_last_error_info()` returns the information for the last
+N-API call that was made and the  `error_code` will match the
+`napi_status` returned by that call.  The `error_message` will
+be a textual representation of the error that occurred.
+
+
+#### napi_get_last_error_info
+##### Signature
+```C
+NAPI_EXTERN napi_status
+napi_get_last_error_info(napi_env env,
+                         const napi_extended_error_info** result);
+```
+##### Parameters
+- `[in]  env `: The environment that the API is invoked under
+- `[out]  result`: The `napi_extended_error_info` structure with more
+information about the error.
+
+##### Return value
+- `napi_ok` if the API succeeded.
+
+##### Description
+This API retrieves a `napi_extended_error_info` structure with information
+about the last error tha occured.
+
+
+### Exceptions
+
+Any N-API function may result in a pending JavaScript exception.  This is
+obviously the case for any function that may cause the execution of JavaScript
+but N-API specifies that an exception may be pending on return from any
+of the API functions.
+
+If the `napi_status` returned by a function is `napi_ok` then no
+exception is pending and no additional action is required.  If the
+`napi_status` returned is anything other than `napi_ok` or
+`napi_pending_exception`, and you want to try to recover and continue
+instead of simply returning immediately, then you must call the
+following function in order to determine if an exception is pending or not:
+
+```C
+napi_is_exception_pending()
+```
+
+and then check the value of `result`.
+
+When an exception is pending you will generally follow one of two approaches.
+
+The first appoach is to do any appropriate cleanup and then return so that
+execution will return to JavaScript.  As part of the transition back to
+JavaScript the exception will be thrown at the point in the JavaScript
+code where the native method was invoked.  The behavior of most N-API calls
+is unspecified while an exception is pending, and many will simply return
+`napi_pending_exception`, so it is important to do as little as possible
+and then return to JavaScript where the exception can be handled.
+
+The second approach is to try to handle the exception. There will be cases
+where the native code can 'catch' the exception, take the appropriate action,
+and then continue.  It is recommended that you only do this in specific cases
+where you know you can safely handle the exception.  In these cases you can
+use the following function to get and clear the exception:
+
+```C
+napi_get_and_clear_last_exception()
+```
+
+On success, result will contain the handle to the JavaScript `Error` object.
```

Comment:
Is this guaranteed to be an `Error` object, or can it be whatever was thrown by JS?